### PR TITLE
Feature/api checker

### PR DESCRIPTION
**Pull request by saxix**

### CODE REVIEW

File: src/etools/libraries/utils/test/api_checker.py [hunk 1]
```diff
@@ -0,0 +1,406 @@
...
+                    if field_name in stored and v != stored[field_name]:
+                        raise AssertionError(rf"""View `{view}` breaks the contract.
+Field `{field_name}` does not match.
+- expected: `{stored[field_name]}`
+- received: `{response[field_name]}`""")
+
+    def _compare(self, response, stored, filename, ignore_fields=None, view='unknown'):
+
+        # ignore_fields = ignore_fields or []
+        # assert isinstance(response, type(stored)), "response and stored do not match"
```
Comment (vkurup) [inline src/etools/libraries/utils/test/api_checker.py]: Are these commented lines needed, and is `ignore_fields` needed?

File: src/etools/libraries/utils/test/api_checker.py [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+New fields are:
+`%s`""" % [f for f in response.keys() if f not in stored.keys()])
+        return True
+
+    def assertAPI(self, url, allow_empty=False, headers=True, status=True):
+        """
+        check url for response changes
+
+        :param url: url to check
+        :param allow_empty: if True ignore empty response and 404 errors
```
Comment (vkurup) [inline src/etools/libraries/utils/test/api_checker.py]: Do you think there's value to allowing each of these cases individually? (i.e. allow_empty_response and allow_404)

File: src/etools/libraries/utils/test/api_checker.py [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+
+        def check_url(url):
+            def _inner(self):
+                self.assertAPI(url)
+
+            _inner.__name__ = "test_url__" + clean_url(u)
+            return _inner
+
+        if 'URLS' not in attributedict:
+            raise ValueError(f"Error creatine {clsname}. "
```
Comment (vkurup) [inline src/etools/libraries/utils/test/api_checker.py]: s/creatine/creating/

File: src/etools/libraries/utils/test/api_checker.py [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+from django.test import Client
+from django.urls import resolve
+
+from adminactions.export import ForeignKeysCollector
+from rest_framework.response import Response
+
+from etools.applications.users.tests.factories import UserFactory
+
+BASE_DATADIR = '_api_checker'
+OVEWRITE = os.environ.get('API_CHECKER_RESET', False)
```
Comment (vkurup) [inline src/etools/libraries/utils/test/api_checker.py]: s/OVEWRITE/OVERWRITE (but also wondering if this feature is really needed. Maybe we should just force the user to delete the old fixture)

File: src/etools/applications/partners/models.py [hunk 5]
```diff
@@ -1319,10 +1319,10 @@ def transition_to_terminated(self):
         pass
 
     @transaction.atomic
-    def save(self, **kwargs):
+    def save(self, force_insert=False, **kwargs):
```
[the remark this answers is not in the document]
Comment (saxix) [reply src/etools/applications/partners/models.py]: force_insert is part of the original method firm. Rely on pk==None to check if the record is a new record is not always true (ie. fixtures load). In the API checker,  records are  always reloaded  with the original id, this is the reason why the fix is required

File: src/etools/libraries/utils/test/api_checker.py [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+
+AssertionError: View `<class 'etools.applications.partners.views.agreements_v2.AgreementListAPIView'>` returned more field than expected.
+Action needed api_v2_agreements.response.json need rebuild.
+New fields are:
+`['country_programme']`
+
+How To use it:
+
+- add ApiCheckerMixin to your Test base classes
+- use ViewSetChecker as your Test metaclass
```
[the remark this answers is not in the document]
Comment (saxix) [reply src/etools/libraries/utils/test/api_checker.py]: yep they are 2 alternatives.

File: src/etools/libraries/utils/test/api_checker.py [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+from adminactions.export import ForeignKeysCollector
+from rest_framework.response import Response
+
+from etools.applications.users.tests.factories import UserFactory
+
+BASE_DATADIR = '_api_checker'
+OVEWRITE = os.environ.get('API_CHECKER_RESET', False)
+
+
+class ResponseEncoder(json.JSONEncoder):
```
[the remark this answers is not in the document]
Comment (saxix) [reply src/etools/libraries/utils/test/api_checker.py]: it described on the top fo the file.
```
This module can also intercept when a field is added,
in this case it is mandatory recreate stored test data; simply delete them from the disk
or set `API_CHECKER_RESET` environment variable and run the test again,
```

File: src/etools/libraries/utils/test/api_checker.py [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+    def default(self, obj):
+        if isinstance(obj, set):
+            return list(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+def mktree(newdir):
+    """works the way a good mkdir should :)
```
[the remark this answers is not in the document]
Comment (saxix) [reply src/etools/libraries/utils/test/api_checker.py]: yep... it is a porting from the old code that in not needed in py3

File: src/etools/applications/partners/models.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (saxix) [reply src/etools/applications/partners/models.py]: @vkurup worked on lib last weekend,  waiting to see if we can use that or remain on this code

File: src/etools/applications/partners/models.py [hunk 9]
```diff
@@ -2024,15 +2024,15 @@ def update_ssfa_properties(self):
...
         # check status auto updates
         # TODO: move this outside of save in the future to properly check transitions
         # self.check_status_auto_updates()
 
         oldself = None
-        if self.pk:
+        if self.pk and not force_insert:
             # load from DB
-            oldself = Intervention.objects.get(pk=self.pk)
+            oldself = Intervention.objects.filter(pk=self.pk).first()
```
Comment (robertavram) [inline src/etools/applications/partners/models.py]: pk should be unique, get should not fail. Why are we changing it?

File: src/etools/libraries/utils/test/api_checker.py [hunk 10]
```diff
@@ -0,0 +1,28 @@
+# -*- coding: utf-8 -*-
+import datetime
+
+from drf_api_checker.unittest import ApiCheckerMixin as _, ApiCheckerBase
+from rest_framework.response import Response
+
+from etools.applications.users.tests.factories import UserFactory
+
+
+class ApiCheckerMixin(_):
```
Comment (robertavram) [inline src/etools/libraries/utils/test/api_checker.py]: global "_" might not be a good idea, can we use a different name for it? as underscore is sometimes used for throwaway variables.

File: src/etools/applications/partners/models.py (hunk 9 above)
Comment (saxix) [reply src/etools/applications/partners/models.py]: `get()` raise ObjectDoesNotExists() where `filter().first()` returns `None`. 
Reading the code we accept that `oldself` is None, and it is always possible that the record does not exists (load fixture/csv import, other process/user deleted the record)  so I think it's safer use this code, but it is not madatory

File: src/etools/libraries/utils/test/api_checker.py (hunk 10 above)
Comment (saxix) [reply src/etools/libraries/utils/test/api_checker.py]: ops... I used that with the intention to fix it later..... and I forgot

File: src/etools/applications/partners/models.py (hunk 9 above)
Comment (robertavram) [reply src/etools/applications/partners/models.py]: if loading from csv import or fixtures with ids should have the force_insert no? I feel like it should explode if you have a pk, no force_insert and there is no record in the DB. Up to you though.. not a major concern.

**Comment by ntrncic** [comment]
@saxix please fix tests.

File: src/requirements/input/test.in [hunk 11]
```diff
@@ -1,4 +1,5 @@
 coverage
+django-adminactions
```
Comment (vkurup) [inline src/requirements/input/test.in]: This requirement was needed for the previous version of this PR, but AFAICT is not needed anymore.

**Pull request merged**